File: book/07-applied-projects/news-aggregator/backend/scripts/01_fetch_one_feed.py
"""
Chapter 1: Foundations
Fetch one RSS feed and print the article titles.

This is deliberately simple — no functions, no classes yet.
We'll refactor this exact logic using FP (Ch3) and OOP (Ch4) later,
so notice how it feels to write it "flat" first.
"""

import urllib.request
import xml.etree.ElementTree as ET

FEED_URL = "https://www.prothomalo.com/feed"  # example Bangla news RSS feed

response = urllib.request.urlopen(FEED_URL)
xml_data = response.read()

root = ET.fromstring(xml_data)

# RSS structure: rss > channel > item > title
for item in root.findall("./channel/item"):
    title = item.find("title").text
    print(title)

File: book/07-applied-projects/news-aggregator/backend/scripts/01_fetch_one_feed.js
/**
 * Chapter 1: Foundations
 * Fetch one RSS feed and print the article titles.
 *
 * Same logic as the Python version, deliberately flat — no functions/classes yet.
 * Run with: node 01_fetch_one_feed.js
 * (requires a DOM-less XML parser like 'fast-xml-parser' — see package.json)
 */

const { XMLParser } = require("fast-xml-parser");

const FEED_URL = "https://www.prothomalo.com/feed"; // example Bangla news RSS feed

async function main() {
  const response = await fetch(FEED_URL);
  const xmlText = await response.text();

  const parser = new XMLParser();
  const data = parser.parse(xmlText);

  const items = data.rss.channel.item;

  for (const item of items) {
    console.log(item.title);
  }
}

main();
